### PR TITLE
IDCOM-1313 Proxy Solana connection for retry

### PR DESCRIPTION
**Pull request by flipscholtz**
- Add a proxy around the Solana connection object, adding retry to the individual `sendTransaction` , `confirmTransaction` etc calls.
- Add checks to the GatekeeperService (one level up) that checks the token status on-chain upon catching a failed tx error. If the token state is already where we want it, it means a previous timed out attempt may have succeeded after all and we don't need to raise an error.
- Add some unit tests for the GatekeeperService error handling logic.

### CODE REVIEW

File: solana/gatekeeper-lib/src/service/GatekeeperService.ts [hunk 1]
```diff
@@ -29,6 +33,35 @@ export type GatekeeperConfig = {
   defaultExpirySeconds?: number;
 };
 
+/**
+ * Enables easier unit testing of gatekeeper-lib as we don't have to stub the imported function from gateway-ts,
+ * we can stub this function which is in the same lib.
+ */
+export const getGatewayTokenKeyForOwner = async (
```
Comment (kevinhcolgan) [inline solana/gatekeeper-lib/src/service/GatekeeperService.ts:40]: not sure I understand how stubbing a local function is very different from stubbing a lib function

File: solana/gatekeeper-lib/src/service/GatekeeperService.ts [hunk 2]
```diff
@@ -100,14 +138,29 @@ export class GatekeeperService {
...
+    } catch (err) {
+      // The tx may have succeeded but the blockchain call timed out. Check on-chain.
+      const onChainToken: GatewayToken | null = await getGatewayToken(
+        this.connection,
+        gatewayTokenKey
+      );
+      if (onChainToken && onChainToken.state === State.ACTIVE) {
+        console.log(
+          "Send failed during Issue, but Active token is found on-chain. Returning success."
+        );
```
Comment (kevinhcolgan) [inline solana/gatekeeper-lib/src/service/GatekeeperService.ts:158]: would it not be better to have the return here and then there's no need for the 'else' statement?

File: solana/gatekeeper-lib/src/service/GatekeeperService.ts [hunk 3]
```diff
@@ -152,14 +205,29 @@ export class GatekeeperService {
...
+        this.payer,
+        this.gatekeeperAuthority
+      );
+    } catch (err) {
+      // The tx may have succeeded but the blockchain call timed out. Check on-chain.
+      const onChainToken: GatewayToken | null = await getGatewayToken(
+        this.connection,
+        gatewayTokenKey
+      );
+      if (onChainToken?.state === State.REVOKED) {
```
Comment (kevinhcolgan) [inline solana/gatekeeper-lib/src/service/GatekeeperService.ts:222]: I think this logic could be abstracted into a function `checkStateOrThrow` or something like that where the token and expected state are passed

Comment (kevinhcolgan) [reply solana/gatekeeper-lib/src/service/GatekeeperService.ts:222]: or even passing a predicate instead of desired state to handle the other cases below

Comment (flipscholtz) [reply solana/gatekeeper-lib/src/service/GatekeeperService.ts]: Done

**Review by kevinhcolgan — APPROVED** [review]

**Pull request merged**